distributed/s3.py

```diff
@@ -27,147 +33,741 @@
-        prefix = prefix[1:]
-
-    L = list(s3.Bucket(bucket_name)
-               .objects.filter(Prefix=prefix, Delimiter=delimiter)
-               .page_size(page_size))
-    return [s for s in L if s.key[-1] != '/']
-
-
-def read_content_from_keys(bucket, key, anon=None):
-    if bucket.startswith('s3://'):
-        bucket = bucket[len('s3://'):]
-    s3 = get_s3(anon)
-    return s3.Object(bucket, key).get()['Body'].read()
-
-
-def read_bytes(bucket_name, prefix='', path_delimiter='', executor=None,
-               lazy=True, anon=None):
-    """ Read data on S3 into bytes in distributed memory
+            if self.start is None and self.end is None:
+                # First read
+                self.start = start
+                self.end = end + self.blocksize
+                self.cache = self.s3.s3.get_object(Bucket=self.bucket, Key=self.key,
+                                                Range='bytes=%i-%i' % (start, self.end - 1)
+                                                )['Body'].read()
+            if start < self.start:
+                new = self.s3.s3.get_object(Bucket=self.bucket, Key=self.key,
+                                         Range='bytes=%i-%i' % (start, self.start - 1)
+                                         )['Body'].read()
+                self.start = start
+                self.cache = new + self.cache
+            if end > self.end:
+                if end > self.size:
+                    return
+                new = self.s3.s3.get_object.get(Bucket=self.bucket, Key=self.key,
+                                             Range='bytes=%i-%i' % (self.end, end + self.blocksize - 1)
+                                             )['Body'].read()
+                self.end = end + self.blocksize
+                self.cache = self.cache + new
+        except ClientError:
+            self.start = min([start, self.start or self.size])
+            self.end = max(end, self.end or self.size)
+
+    def read(self, length=-1):
+        """
+        Return data from cache, or fetch pieces as necessary
+        """
+        if self.mode != 'rb':
+            raise ValueError('File not in read mode')
+        if length < 0:
+            length = self.size
+        if self.closed:
+            raise ValueError('I/O operation on closed file.')
+        self._fetch(self.loc, self.loc + length)
+        out = self.cache[self.loc - self.start:
+                         self.loc - self.start + length]
+        self.loc += len(out)
+        return out
+
+    def write(self, data):
+        if self.mode != 'wb':
+            raise ValueError('File not in write mode')
+        partno = len(self.part_info['Parts']) + 1
+        part = self.s3.s3.upload_part(Bucket=self.bucket, Key=self.key,
+                                      PartNumber=partno,
+                                      UploadId=self.mpu['UploadId'], Body=data)
+        self.part_info['Parts'].append({'PartNumber': partno, 'ETag': part['ETag']})
```

Comment:
We'll have to either test or remove this extra functionality before merging.  What would be best?